docs/specifying_parameters.md
# Specifying Parameters

Running a single experiment with Zeno Build can test many different models
and parameter settings. Exactly which parameters are tested is specified in
the `config.py` file in each example directory.

Here we will provide an example using for the [chatbot](/examples/chatbot/)
task.

## Search Space

The most important parameter is the `search_space` parameter, which specifies
the search space of hyperparameters to test. This is specified in the
`config.py` file as follows:

```python
space = search_space.CombinatorialSearchSpace(
    {
        "dataset_preset": search_space.Constant("dstc11"),
        "model_preset": search_space.Categorical(
            ["gpt2-xl", "llama-7b", "vicuna-7b", "mpt-7b-chat"]
        ),
        "prompt_preset": search_space.Categorical(
            ["standard", "friendly", "polite", "cynical"]
        ),
        "temperature": search_space.Float(0.2, 0.4),
        "context_length": search_space.Discrete([1, 2, 3, 4]),
        "max_tokens": search_space.Constant(100),
        "top_p": search_space.Constant(1.0),
    }
)
```

Here the `search_space` parameter is a `CombinatorialSearchSpace` object,
which indicates that we should search over the cross product of all the
parameters specified in the dictionary. Each parameter in the dictionary
can be of a variety of types:

- `search_space.Constant`: A constant value that is not searched over.
- `search_space.Categorical`: A categorical value, such as a list of strings.
- `search_space.Discrete`: A discrete value, such as a list of integers.
  The difference between `Categorical` and `Discrete`
  is that there is an inherent ordering in `Discrete` values, but not
  in `Categorical` values.
- `search_space.Float`: A float value, such as a float between 0 and 1.
- `search_space.Int`: An integer value, such as an integer between 0 and 10.

## Number of Trials

There is also a `num_trials` parameter, which specifies the number of unique
parameter settings to try out.

```python
num_trials = 10
```

## Data, Model, Prompt Presets

Each task is slightly different, so the parameters that are searched over
may be different. But most tasks can have a `dataset_preset` and
`model_preset` parameter to specify the data and the model. And when using
a language model by providing it with a prompt, a `prompt_preset` parameter
can be used to specify the prompt.

### Data Configurations

Data configurations can specify the dataset name (usually from
[Hugging Face Datasets](https://huggingface.co/datasets)), the split of
that dataset, that column in the dataset to use (and also the `label_column`
if we have a separate column for labels), and the format of the data.
Here are some examples for chatbots:

```python
dataset_configs = {
    "dstc11": DatasetConfig(
        dataset="gneubig/dstc11",
        split="validation",
        data_column="turns",
        data_format="dstc11",
    ),
    "daily_dialog": DatasetConfig(
        dataset="daily_dialog",
        split="validation",
        data_column="dialog",
        data_format="sequence",
    ),
}
```

### Model Configurations

Model configurations can specify the provider of the model, such as OpenAI,
Cohere, or Hugging Face, the name of the model. In some cases extra auxiliary
parameters such as the model class or tokenizer class may be necessary.

```python
model_configs = {
    "text-davinci-003": LMConfig(provider="openai", model="text-davinci-003"),
    "gpt-3.5-turbo": LMConfig(provider="openai_chat", model="gpt-3.5-turbo"),
    "cohere-command-xlarge": LMConfig(
        provider="cohere", model="command-xlarge-nightly"
    ),
    "gpt2": LMConfig(
        provider="huggingface",
        model="gpt2",
        model_cls=transformers.GPT2LMHeadModel,
    ),
    "llama-7b": LMConfig(
        provider="huggingface",
        model="decapoda-research/llama-7b-hf",
        tokenizer_cls=transformers.LlamaTokenizer,
    ),
    ...
}
```

### Prompt Configurations

Prompt configurations can specify the type of prompt to use. Here are
a few examples:

```python
prompt_messages: dict[str, ChatMessages] = {
    "standard": ChatMessages(
        messages=[
            ChatTurn(
                role="system",
                content="You are a chatbot tasked with making small-talk with "
                "people.",
            ),
        ]
    ),
    "friendly": ChatMessages(
        messages=[
            ChatTurn(
                role="system",
                content="You are a kind and friendly chatbot tasked with making "
                "small-talk with people in a way that makes them feel "
                "pleasant.",
            ),
        ]
    ),
    ...
}
```

## Distill and Metric Functions

Finally, we define `distill` and `metric` functions. `distill` functions are
essentially [Zeno](https://zenoml.com)'s terminology for "feature" functions,
that calculate a feature
vector for a given input. `metric` functions are functions that calculate an
evaluation metric for an entire dataset, sometimes taking the output of a
`distill` function as input.
Zeno Build provides a library of various distill and metric functions
in the [evaluation](../zeno_build/evaluation/) directory, so you can
browse through the full library there.

We use distill/metric functions at two points in
the experimentation process.

First, we use them while making a sweep over the various parameter values,
evaluating a metric value after the end of every experiment. Here is an
example of using the [chrf](https://aclanthology.org/W15-3049/) string
overlap metric. We specify both the distill function `chrf`, which calculates
the `chrf` for each example, and the metric function `avg_chrf`, which
calculates the average `chrf` for the entire dataset.

```python
sweep_distill_functions = [chrf]
sweep_metric_function = avg_chrf
```

Second, we use them in Zeno visualization for
[exploring results](exploring_results.md), specifying both

- Distill functions used to slice and dice the data such as `output_length`,
  `input_length`, `label_length`, `chat_context_length`.
- Distill functions used to calculate metrics such as `chrf`, `length_ratio`,
  `bert_score`, `exact_match`.
- Metric functions used to calculate average metrics such as `avg_chrf`,
    `avg_length_ratio`, `avg_bert_score`, `avg_exact_match`.

```python
zeno_distill_and_metric_functions = [
    output_length,
    input_length,
    label_length,
    chat_context_length,
    chrf,
    length_ratio,
    bert_score,
    exact_match,
    avg_chrf,
    avg_length_ratio,
    avg_bert_score,
    avg_exact_match,
]
```

## Next Steps

Now that we've implemented our model and defined the hyperparameters,
we can move on to [running experiments](running_experiments.md).


examples/chatbot/config.py
"""Various configuration options for the chatbot task.

This file is intended to be modified. You can go in and change any
of the variables to run different experiments.
"""

from __future__ import annotations

import transformers

from zeno_build.evaluation.text_features.clustering import label_clusters
from zeno_build.evaluation.text_features.exact_match import avg_exact_match, exact_match
from zeno_build.evaluation.text_features.length import (
    chat_context_length,
    input_length,
    label_length,
    output_length,
)
from zeno_build.evaluation.text_features.numbers import english_number_count
from zeno_build.evaluation.text_metrics.critique import (
    avg_bert_score,
    avg_chrf,
    avg_length_ratio,
    bert_score,
    chrf,
    length_ratio,
)
from zeno_build.experiments import search_space
from zeno_build.models.dataset_config import DatasetConfig
from zeno_build.models.lm_config import LMConfig
from zeno_build.prompts.chat_prompt import ChatMessages, ChatTurn

# --- Model Configuration ---

# The details of each model
model_configs = {
    "text-davinci-003": LMConfig(provider="openai", model="text-davinci-003"),
    "gpt-3.5-turbo": LMConfig(provider="openai_chat", model="gpt-3.5-turbo"),
    "cohere-command-xlarge": LMConfig(
        provider="cohere", model="command-xlarge-nightly"
    ),
    "gpt2": LMConfig(
        provider="huggingface",
        model="gpt2",
    ),
    "gpt2-xl": LMConfig(
        provider="huggingface",
        model="gpt2-xl",
    ),
    # We need to use the transformers library instead of VLLM here
    # because the tokenizer library needs to be set manually
    "llama-7b": LMConfig(
        provider="huggingface",
        model="decapoda-research/llama-7b-hf",
        tokenizer_cls=transformers.LlamaTokenizer,
    ),
    "llama-13b": LMConfig(
        provider="huggingface",
        model="decapoda-research/llama-13b-hf",
        tokenizer_cls=transformers.LlamaTokenizer,
    ),
    "vicuna-7b": LMConfig(
        provider="huggingface",
        model="eachadea/vicuna-7b-1.1",
        name_replacements={
            "system": "ASSISTANT",
            "assistant": "ASSISTANT",
            "user": "HUMAN",
        },
    ),
    "vicuna-13b": LMConfig(
        provider="huggingface",
        model="eachadea/vicuna-13b-1.1",
        name_replacements={
            "system": "ASSISTANT",
            "assistant": "ASSISTANT",
            "user": "HUMAN",
        },
    ),
    "vicuna-7b-v1.3": LMConfig(
        provider="huggingface",
        model="lmsys/vicuna-7b-v1.3",
        name_replacements={
            "system": "ASSISTANT",
            "assistant": "ASSISTANT",
            "user": "HUMAN",
        },
    ),
    "vicuna-13b-v1.3": LMConfig(
        provider="huggingface",
        model="lmsys/vicuna-13b-v1.3",
        name_replacements={
            "system": "ASSISTANT",
            "assistant": "ASSISTANT",
            "user": "HUMAN",
        },
    ),
    "vicuna-33b-v1.3": LMConfig(
        provider="huggingface",
        model="lmsys/vicuna-33b-v1.3",
        name_replacements={
            "system": "ASSISTANT",
            "assistant": "ASSISTANT",
            "user": "HUMAN",
        },
    ),
    # We need to use huggingface instead of vllm here because we need to
    # set trust_remote_code to True
    "mpt-7b-chat": LMConfig(
        provider="huggingface",
        model="mosaicml/mpt-7b-chat",
        model_loader_kwargs={"trust_remote_code": True},
    ),
}

# These models are used by default in the experiments.
# This can be modified by using the "--models" command line argument.
default_models = [
    "gpt-3.5-turbo",
    "gpt2",
    "gpt2-xl",
    "llama-7b",
    "vicuna-7b",
    "mpt-7b-chat",
]
# The default single model to use in experiments that don't iterate over
# multiple models.
default_single_model = "vicuna-7b"

# --- Dataset Configuration ---

# The details of each dataset
dataset_configs = {
    "dstc11": DatasetConfig(
        dataset="gneubig/dstc11",
        split="validation",
        data_column="turns",
        data_format="dstc11",
    ),
}

# --- Prompt Configuration ---

# The details of the prompts
prompt_messages: dict[str, ChatMessages] = {
    "standard": ChatMessages(
        messages=[
            ChatTurn(
                role="system",
                content="You are a chatbot tasked with making small-talk with "
                "people.",
            ),
        ]
    ),
    "friendly": ChatMessages(
        messages=[
            ChatTurn(
                role="system",
                content="You are a kind and friendly chatbot tasked with making "
                "small-talk with people in a way that makes them feel "
                "pleasant.",
            ),
        ]
    ),
    "polite": ChatMessages(
        messages=[
            ChatTurn(
                role="system",
                content="You are an exceedingly polite chatbot that speaks very "
                "formally and tries to not make any missteps in your "
                "responses.",
            ),
        ]
    ),
    "cynical": ChatMessages(
        messages=[
            ChatTurn(
                role="system",
                content="You are a cynical chatbot that has a very dark view of the "
                "world and in general likes to point out any possible "
                "problems.",
            ),
        ]
    ),
    # The following is purpose-tailored for the DSTC11 insurance dataset
    "insurance_standard": ChatMessages(
        messages=[
            ChatTurn(
                role="system",
                content="You are an agent at the Rivertown Insurance helpdesk that "
                "mainly helps with resolving insurance claims.",
            ),
        ]
    ),
    # The following is purpose-tailored for the DSTC11 insurance dataset
    "insurance_upgrade_1": ChatMessages(
        messages=[
            ChatTurn(
                role="system",
                content="""\n
You are an agent at the Rivertown Insurance helpdesk that helps with resolving insurance
claims.

Make sure you introduce yourself appropriately, example:
> Assistant: Hello. Thank you for calling Rivertown Insurance. How can I help you?

When people provide numbers like their security number, make sure that you repeat the
number back to them to confirm that you have the correct number, example:
> User: Is the account number eight digit or ten digit?
> Assistant: It is eight digit.
> User: Okay. Four five.
> Assistant: Four five.""",
            ),
        ]
    ),
}

default_prompts = list(prompt_messages.keys())
# The default prompt to use in experiments that don't iterate over
# multiple prompts.
default_single_prompt = "standard"

# --- Other Hyperparameters ---

default_temperatures = [0.2, 0.3, 0.4]
default_single_temperature = 0.3

default_context_lengths = [1, 2, 3, 4, 6, 8]
default_single_context_length = 4

default_single_max_tokens = 100
default_single_max_p = 1.0

dataset = "dstc11"

# --- Evaluation/Feature Configuartion ---

# The functions to use to calculate scores for the hyperparameter sweep
sweep_distill_functions = [chrf]
sweep_metric_function = avg_chrf

# The functions used for Zeno visualization
zeno_distill_and_metric_functions = [
    output_length,
    input_length,
    label_length,
    chat_context_length,
    english_number_count,
    label_clusters,
    chrf,
    length_ratio,
    bert_score,
    exact_match,
    avg_chrf,
    avg_length_ratio,
    avg_bert_score,
    avg_exact_match,
]

# --- Experiment Configuration ---

# A bunch of different experiments that could be run. Which ones to run
# is controlled by the "--experiments" command line argument.
experiments = {
    # An exhaustive experiment that tests many different combinations
    "exhaustive": search_space.CombinatorialSearchSpace(
        {
            "model_preset": search_space.Categorical(default_models),
            "prompt_preset": search_space.Categorical(default_prompts),
            "temperature": search_space.Discrete(default_temperatures),
            "context_length": search_space.Discrete(default_context_lengths),
            "max_tokens": search_space.Constant(default_single_max_tokens),
            "top_p": search_space.Constant(default_single_max_p),
        }
    ),
    # An experiment that varies only the model
    "model": search_space.CombinatorialSearchSpace(
        {
            "model_preset": search_space.Categorical(default_models),
            "prompt_preset": search_space.Constant(default_single_prompt),
            "temperature": search_space.Constant(default_single_temperature),
            "context_length": search_space.Constant(default_single_context_length),
            "max_tokens": search_space.Constant(default_single_max_tokens),
            "top_p": search_space.Constant(default_single_max_p),
        }
    ),
    # An experiment that varies only the prompt
    "prompt": search_space.CombinatorialSearchSpace(
        {
            "model_preset": search_space.Constant(default_single_model),
            "prompt_preset": search_space.Categorical(default_prompts),
            "temperature": search_space.Constant(default_single_temperature),
            "context_length": search_space.Constant(default_single_context_length),
            "max_tokens": search_space.Constant(default_single_max_tokens),
            "top_p": search_space.Constant(default_single_max_p),
        }
    ),
    # An experiment that varies only the temperature
    "temperature": search_space.CombinatorialSearchSpace(
        {
            "model_preset": search_space.Constant(default_single_model),
            "prompt_preset": search_space.Constant(default_single_prompt),
            "temperature": search_space.Discrete(default_temperatures),
            "context_length": search_space.Constant(default_single_context_length),
            "max_tokens": search_space.Constant(default_single_max_tokens),
            "top_p": search_space.Constant(default_single_max_p),
        }
    ),
    # An experiment that varies only the context_length
    "context_length": search_space.CombinatorialSearchSpace(
        {
            "model_preset": search_space.Constant(default_single_model),
            "prompt_preset": search_space.Constant(default_single_prompt),
            "temperature": search_space.Constant(default_single_temperature),
            "context_length": search_space.Discrete(default_context_lengths),
            "max_tokens": search_space.Constant(default_single_max_tokens),
            "top_p": search_space.Constant(default_single_max_p),
        }
    ),
}

# The number of trials to run. If set to None, all combinations of experiments will be
# run.
num_trials: int | None = None
